Documentation/format/index.md

```diff
@@ -0,0 +1,198 @@
+│ │                  │ ┌──────────────────────────┐     │ │
+│ │                  │ │ ref(l_i.name) <varint>   │     │ │
+│ │ #labels <varint> │ ├──────────────────────────┤ ... │ │
+│ │                  │ │ ref(l_i.value) <varint>  │     │ │
+│ │                  │ └──────────────────────────┘     │ │
+│ ├──────────────────┼──────────────────────────────────┤ │
+│ │                  │ ┌──────────────────────────┐     │ │
+│ │                  │ │ c_i.mint <varint>        │     │ │
+│ │                  │ ├──────────────────────────┤     │ │
+│ │                  │ │ c_i.maxt <varint>        │     │ │
+│ │ #chunks <varint> │ ├──────────────────────────┤ ... │ │
+│ │                  │ │ ref(c_i.data) <varint>   │     │ │
+│ │                  │ ├──────────────────────────┤     │ │
+│ │                  │ │ crc32(c_i.data) <varint> │     │ │
+│ │                  │ └──────────────────────────┘     │ │
+│ └──────────────────┴──────────────────────────────────┘ │
+├─────────────────────────────────────────────────────────┤
+│ CRC32 <4 byte>                                          │
+└─────────────────────────────────────────────────────────┘
+```
+
+The CRC checksum is calculated over the series contents of the index concatenated with the data of its chunks (with encoding byte, without length).
+
+
+### Label Index
+
+The label index indexes holds lists of possible values for label names. A sequence of label index blocks follow on the series entries.
+
+```
+┌─────────────────────────────────────────────────────────┐
+│ len <varint>                                            │
+├─────────────────────────────────────────────────────────┤
+│ ┌──────────────────┬──────────────────────────────────┐ │
+│ │                  │ ┌──────────────────────────┐     │ │
+│ │                  │ │ ref(value[0]) <4 byte>   │     │ │
+│ │                  │ ├──────────────────────────┤     │ │
+│ │  n = len(names)  │ │ ...                      │ ... │ │
+│ │     <varint>     │ ├──────────────────────────┤     │ │
+│ │                  │ │ ref(value[n]) <4 byte>   │     │ │
+│ │                  │ └──────────────────────────┘     │ │
+│ └──────────────────┴──────────────────────────────────┘ │
+├─────────────────────────────────────────────────────────┤
+│ CRC32 <4 byte>                                          │
+└─────────────────────────────────────────────────────────┘
+```
+
+The sequence of label index blocks is finalized by a lookup table pointing to the beginning of each label index block. It is simply a list of entries that are read into an in-memory hashmap when the index is loaded.
+
+### Postings
+
+Postings are postings lists that map label pairs to series they occur in.
+
+```
+┌─────────────────────────────────────────────────┐
+│ len <varint>                                    │
+├─────────────────────────────────────────────────┤
+│ ┌─────────────────────────────────────────────┐ │
+│ │ ref(series[0]) <4 byte>                     │ │
```

Comment:
Considering the options where and how to put that padding. I think it's best to make it concern of the serializer and not of the format, i.e. don't make padding part of the entry itself.

The writer can just add padding before and the reference will be correct. This makes sequential reading of all postings not work directly anymore. But the simple rule that 0 bytes should just be skipped until the next entries' header, which must never start with 0, is encountered seems to address this just fine.

This should generally allow us to work with inter-entry paddings without bumping the format.